java-support/src/main/scala/io/cloudstate/javasupport/impl/crdt/AnnotationBasedCrdtExtensionSupport.scala

```diff
@@ -0,0 +1,118 @@
+        val name: String = if (annotation.name().isEmpty) {
+          ReflectionHelper.getCapitalizedName(method)
+        } else annotation.name()
+
+        val serviceMethod = resolvedMethods.getOrElse(name, {
+          throw new RuntimeException(
+            s"Command handler method ${method.getName} for command $name found, but the service has no command by that name."
+          )
+        })
+        (ReflectionHelper.ensureAccessible(method), serviceMethod)
+      }
+
+    def getHandlers[C <: CrdtContext with CrdtFactory: ClassTag](streamed: Boolean) =
+      handlers
+        .filter(_._2.outputStreamed == streamed)
+        .map {
+          case (method, serviceMethod) =>
+            new CommandHandlerInvoker[C](method, serviceMethod, CrdtAnnotationHelper.crdtParameterHandlers[C])
+        }
+        .groupBy(_.serviceMethod.name)
+        .map {
+          case (commandName, Seq(invoker)) => commandName -> invoker
+          case (commandName, many) =>
+            throw new RuntimeException(
+              s"Multiple methods found for handling command of name $commandName: ${many.map(_.method.getName)}"
+            )
+        }
+
+    (getHandlers[CommandContext](false), getHandlers[StreamedCommandContext[AnyRef]](true))
+  }
+
+  // TODO JavaDoc
+  override def create(context: CrdtCreationContext): CrdtEntityHandler = {
+    val entity = entitySupportFactory.create(context, null)
+    new EntityHandler(entity)
+  }
+
+  private class EntityHandler(entity: AnyRef) extends CrdtEntityHandler {
+
+    override def handleCommand(command: JavaPbAny, context: CommandContext): Optional[JavaPbAny] = unwrap {
+      val maybeResult = commandHandlers.get(context.commandName()).map { handler =>
+        handler.invoke(entity, command, context)
+      }
+
+      maybeResult.getOrElse {
+        throw new RuntimeException(
+          s"No command handler found for command [${context.commandName()}] on CRDT entity: $entityClass"
+        )
+      }
+    }
+
+    override def handleStreamedCommand(command: JavaPbAny,
+                                       context: StreamedCommandContext[JavaPbAny]): Optional[JavaPbAny] = unwrap {
+      val maybeResult = streamedCommandHandlers.get(context.commandName()).map { handler =>
+        val adaptedContext =
+          new AdaptedStreamedCommandContext(context,
+                                            handler.serviceMethod.outputType.asInstanceOf[ResolvedType[AnyRef]])
+        handler.invoke(entity, command, adaptedContext)
+      }
+
+      maybeResult.getOrElse {
+        throw new RuntimeException(
+          s"No streamed command handler found for command [${context.commandName()}] on CRDT entity: $entityClass"
+        )
+      }
+    }
+
+    private def unwrap[T](block: => T): T =
+      try {
+        block
+      } catch {
+        case ite: InvocationTargetException if ite.getCause != null =>
+          throw ite.getCause
+      }
+  }
```

Comment:
Might be cleaner with

```suggestion
    val andUnwrap: PartialFunction[Throwable, Nothing] = {
        case ite: InvocationTargetException if ite.getCause != null =>
          throw ite.getCause
      }
```

and then instead of:

`unwrap { … }`

you can do:

` try { … } catch andUnwrap`